src/main/java/it/gov/pagopa/pu/sil/service/AlgorithmResolverService.java

```diff
@@ -0,0 +1,45 @@
+package it.gov.pagopa.pu.sil.service;
+
+import com.auth0.jwt.algorithms.Algorithm;
+import it.gov.pagopa.pu.organization.dto.generated.JwtAlgorithm;
+import it.gov.pagopa.pu.sil.util.CertUtils;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+@Slf4j
+@Component
+public class AlgorithmResolverService {
+  private static final String RSA = "RSA";
+  private static final String ECDSA = "EC";
+
+  private final ConcurrentMap<JwtAlgorithm, Algorithm> algorithmCache = new ConcurrentHashMap<>();
```

Comment:
you cannot cache based on the algorithm! you will have different Algorithm based on the key...
```suggestion
  private final ConcurrentMap<Pair<JwtAlgorithm, String>, Algorithm> algorithmCache = new ConcurrentHashMap<>();
```